### PR TITLE
scopes: arguments, objects and array inner functions 

### LINKED ISSUE #188: False positives in no-const-assign
Trying to run latest code on `std/` in `deno` repository I get:
```
(no-const-assign) Reassigning constant variable is not allowed
  actual = await buf.peek(1);
  ~~~~~~~~~~~~~~~~~~~~~~~~~~
    at /Users/biwanczuk/dev/deno/test_util/../std/io/bufio_test.ts:274:2

(no-const-assign) Reassigning constant variable is not allowed
  actual = await buf.peek(1);
  ~~~~~~~~~~~~~~~~~~~~~~~~~~
    at /Users/biwanczuk/dev/deno/test_util/../std/io/bufio_test.ts:278:2

(no-const-assign) Reassigning constant variable is not allowed
  actual = await buf.peek(1);
  ~~~~~~~~~~~~~~~~~~~~~~~~~~
    at /Users/biwanczuk/dev/deno/test_util/../std/io/bufio_test.ts:282:2

(no-const-assign) Reassigning constant variable is not allowed
  actual = await buf.peek(2);
  ~~~~~~~~~~~~~~~~~~~~~~~~~~
    at /Users/biwanczuk/dev/deno/test_util/../std/io/bufio_test.ts:286:2

(no-const-assign) Reassigning constant variable is not allowed
  actual = await buf.peek(4);
  ~~~~~~~~~~~~~~~~~~~~~~~~~~
    at /Users/biwanczuk/dev/deno/test_util/../std/io/bufio_test.ts:294:2

(no-const-assign) Reassigning constant variable is not allowed
  actual = await buf.peek(0);
  ~~~~~~~~~~~~~~~~~~~~~~~~~~
    at /Users/biwanczuk/dev/deno/test_util/../std/io/bufio_test.ts:301:2

(no-const-assign) Reassigning constant variable is not allowed
    eventsFired = [];
    ~~~~~~~~~~~~~~~~
    at /Users/biwanczuk/dev/deno/test_util/../std/node/events_test.ts:28:4
```

CC @disizali

### PR DESCRIPTION
**Pull request by disizali**
fixes #188.
On the way to completing #160.

object/array literal inner `Function` ,`Method`,`Getter`,`Setter` scopes as following :
```ts
let obj = {
  get getter(){} // Getter Scope
  set setter(){} // Setter Scope
  method(){} // Method Scope
}

let array = [
  function inArrayFunction(){} // Function scope
]
```

support `object prop call` and `new object` functions args as following :
```ts
Deno.test("arg test" , function(){}) // Function Scope
new Something(function(){}); // Function Scope
```

### CODE REVIEW

File: src/scopes.rs [hunk 1]
```diff
@@ -35,12 +35,15 @@ pub enum ScopeKind {
...
   Function,
+  Method,
   Block,
   Loop,
   Class,
   Switch,
   With,
   Catch,
+  Getter,
+  Setter,
```
Comment (bartlomieju) [inline src/scopes.rs:46]: Do we really need new scope types? Shouldn't `ScopeKind::Function` be enough?

File: src/scopes.rs [hunk 2]
```diff
@@ -772,6 +930,198 @@ switch (foo) {
...
+      ()=>{const b;},
+      [
+        function nested() { const c;}
+      ],
+      {
+        innerMethod(){
+          const d;
+        }
+      }
+    ]
```
Comment (bartlomieju) [inline src/scopes.rs:1070]: Nice 👍

File: src/scopes.rs (hunk 1 above)
Comment (disizali) [reply src/scopes.rs]: your right, **we** don't need them for the current rules, actually.
Done 👍 
stay minimal for now.


_I'm just curious about doing something in the future, Shouldn't we care about plugin development and analyzing all kinds of scopes? My opinion is that we should proceed according to the needs_

Comment (bartlomieju) [reply src/scopes.rs]: > I'm just curious about doing something in the future, Shouldn't we care about plugin development and analyzing all kinds of scopes? 

I'm not sure - ESLint doesn't recognize `Getter/Setter` scopes so maybe it's critical

> My opinion is that we should proceed according to the needs

I agree

**Pull request merged**